taichi/util/hash.h

```diff
@@ -0,0 +1,33 @@
+#pragma once
+
+#include <functional>
+#include <stddef.h>
+
+namespace taichi::hashing {
+
+namespace {
+inline void hash_combine(size_t &seed, size_t value) {
+  seed ^= value + 0x9e3779b9 + (seed << 6) + (seed >> 2);
```

Comment:
Since we're implementing a custom hash algorithm, can you comment on how did you decide the math model and the parameters "0x9e3779b9", "<< 6" and ">> 2" ? 

Btw, do you also have a rough estimation of the collision rate?

Reply:
It's borrowed from boost, I'll add an comment about the reference.